File: split/getstringsstarted.pas
function getstringsstarted: boolean;
label
  30, 10;
var
  k, l: 0..255;
  m, n: char;
  g: strnumber;
  a: integer;
  c: boolean;
begin
  poolptr := 0;
  strptr := 0;
  strstart[0] := 0;
  {48:}
for k := 0 to 255 do
  begin
    if ({49:}
(k < 32) or (k > 126){:49}
) then
    begin
      begin
        strpool[poolptr] := 94;
        poolptr := poolptr + 1;
      end;
      begin
        strpool[poolptr] := 94;
        poolptr := poolptr + 1;
      end;
      if k < 64 then
      begin
        strpool[poolptr] := k + 64;
        poolptr := poolptr + 1;
      end
      else if k < 128 then
      begin
        strpool[poolptr] := k - 64;
        poolptr := poolptr + 1;
      end
      else
      begin
        l := k div 16;
        if l < 10 then
        begin
          strpool[poolptr] := l + 48;
          poolptr := poolptr + 1;
        end
        else
        begin
          strpool[poolptr] := l + 87;
          poolptr := poolptr + 1;
        end;
        l := k mod 16;
        if l < 10 then
        begin
          strpool[poolptr] := l + 48;
          poolptr := poolptr + 1;
        end
        else
        begin
          strpool[poolptr] := l + 87;
          poolptr := poolptr + 1;
        end;
      end;
    end
    else
    begin
      strpool[poolptr] := k;
      poolptr := poolptr + 1;
    end;
    g := makestring;
  end{:48}
;
  {51:}
nameoffile := poolname;
  if aopenin(poolfile) then
  begin
    c := false;
    repeat{52:}

      begin
        if EOF(poolfile) then
        begin
          ;
          writeln(termout, '! TEX.POOL has no check sum.');
          aclose(poolfile);
          getstringsstarted := false;
          goto 10;
        end;
        read(poolfile, m, n);
        if m = '*' then{53:}

        begin
          a := 0;
          k := 1;
          while true do
          begin
            if (xord[n] < 48) or (xord[n] > 57) then
            begin
              ;
              writeln(termout, '! TEX.POOL check sum doesn''t have nine digits.');
              aclose(poolfile);
              getstringsstarted := false;
              goto 10;
            end;
            a := 10 * a + xord[n] - 48;
            if k = 9 then
              goto 30;
            k := k + 1;
            read(poolfile, n);
          end;
          30:
            if a <> 117275187 then
            begin
              ;
              writeln(termout, '! TEX.POOL doesn''t match; TANGLE me again.');
              aclose(poolfile);
              getstringsstarted := false;
              goto 10;
            end;
          c := true;
        end{:53}

        else
        begin
          if (xord[m] < 48) or (xord[m] > 57) or (xord[n] < 48) or (xord[n] > 57) then
          begin
            ;
            writeln(termout, '! TEX.POOL line doesn''t begin with two digits.');
            aclose(poolfile);
            getstringsstarted := false;
            goto 10;
          end;
          l := xord[m] * 10 + xord[n] - 48 * 11;
          if poolptr + l + stringvacancies > poolsize then
          begin
            ;
            writeln(termout, '! You have to increase POOLSIZE.');
            aclose(poolfile);
            getstringsstarted := false;
            goto 10;
          end;
          for k := 1 to l do
          begin
            if eoln(poolfile) then
              m := ' '
            else
              read(poolfile, m);
            begin
              strpool[poolptr] := xord[m];
              poolptr := poolptr + 1;
            end;
          end;
          readln(poolfile);
          g := makestring;
        end;
      end{:52}
;
    until c;
    aclose(poolfile);
    getstringsstarted := true;
  end
  else
  begin
    ;
    writeln(termout, '! I can''t read TEX.POOL.');
    aclose(poolfile);
    getstringsstarted := false;
    goto 10;
  end{:51}
;
  10: ;
end;

File: cpp/getstringsstarted.cpp
#include "getstringsstarted.h"
#include "makestring.h"
#include "aopenin.h"
#include "aclose.h"
#include <iostream> 

bool getstringsstarted(void)
{
	poolptr = 0;
	strptr = 0;
	strstart[0] = 0;
	for (int k = 0; k < 256; k++)
	{
		if (k < 32 || k > 126)
		{
			strpool[poolptr++] = '^';
			strpool[poolptr++] = '^';
			if (k < 64)
				strpool[poolptr++] = k+64;
			else 
				if (k < 128)
					strpool[poolptr++] = k-64;
				else
				{
					int l = k/16;
					strpool[poolptr++] = l < 10 ? l+'0' : l+'a'-10;
					l = k%16;
					strpool[poolptr++] = l < 10 ? l+'0' : l+'a'-10;
				}
		}
		else
			strpool[poolptr++] = k;
		g = makestring();
	}
	nameoffile = poolname;
	if (aopenin(poolfile))
	{
		c = false;
		do
		{
			if (poolfile.eof())
			{
				std::cout << "! TEX.POOL has no check sum.'" << std::endl;
				aclose(poolfile);
				return false;
			}
			char m = poolfile.get();
			char n = poolfile.get();
			if (m == '*') // checksum
			{
				int a = 0;
				int k = 1;
				while (true)
				{
					if (xord[n] < '0' || xord[n] > '9')
					{
						std::cout << "! TEX.POOL check sum doesn''t have nine digits." << std::endl;
						aclose(poolfile);
						return false;
					}
					a = 10*a+xord[n]-'0';
					if (k == 9)
						break;
					k++;
					n = poolfile.get();
				}
				if (a != 117275187)
				{
					std::cout << "! TEX.POOL doesn''t match; TANGLE me again." << std::endl;
					aclose(poolfile);
					return false;
				}
				c = true;
			}
			else
			{
				if (xord[m] < '0' || xord[m] > '9' || xord[n] < '0' || xord[n] > '9')
				{
					std::cout << "! TEX.POOL line doesn''t begin with two digits." << std::endl;
					aclose(poolfile);
					return false;
				}
				int l = (xord[m]-'0')*10+xord[n]-'0';
				if (poolptr+l+stringvacancies > poolsize)
				{
					std::cout << "! You have to increase POOLSIZE." << std::endl;
					aclose(poolfile);
					return false;
				}
				for (int k = 1; k <= l; k++)
				{
					if (poolfile.peek() != '\n')
						m = ' ';
					else
						m = poolfile.get();
					strpool[poolptr++] = xord[m];
				}
				std::string dummy;
				getline(poolfile, dummy);
				g = makestring();
			}
		} while (!c);
		aclose(poolfile);
		return true;
	}
	else
	{
		std::cout << "! I can''t read TEX.POOL." << std::endl;
		aclose(poolfile);
		return false;
	}
}

